qiling/debugger/qdb/render/render.py

```diff
@@ -3,237 +3,258 @@
-                self.print_asm(line)
+        # number of hexadecimal nibbles to display per value
+        nibbles = self.pointersize * 2
 
-        # assembly for current location
-        if (cur_insn := lines.get("current", None)):
-            prophecy = self.predictor.predict()
-            self.print_asm(cur_insn, to_jump=prophecy.going)
+        for address, value, deref in dump:
+            offset = address - sp
 
-        # assembly after current location
-        if (forward := lines.get("forward", None)):
-            for line in forward:
-                self.print_asm(line)
+            value_str = '(unreachable)' if value is None else f'{value:#0{nibbles + 2}x}'
 
-    def swap_reg_name(self, cur_regs: Mapping[str, int], extra_dict=None) -> Mapping[str, int]:
-        """
-        swap register name with more readable register name
-        """
+            if isinstance(deref, int):
+                deref_str = f'{deref:#0{nibbles + 2}x}'
 
-        target_items = extra_dict.items() if extra_dict else self.regs_need_swapped.items()
+            elif isinstance(deref, str):
+                deref_str = f'"{deref}"'
 
-        for old_reg, new_reg in target_items:
-            cur_regs.update({old_reg: cur_regs.pop(new_reg)})
+            else:
+                deref_str = ''
 
-        return cur_regs 
+            print(f'SP + {offset:#04x} │ {address:#010x} : {value_str}{f" {RARROW} {deref_str}" if deref_str else ""}')
 
-    def print_asm(self, insn: CsInsn, to_jump: bool = False) -> None:
-        """
-        helper function for printing assembly instructions, indicates where we are and the branch prediction
-        provided by BranchPredictor
+    def render_assembly(self, listing: Sequence[InsnLike], pc: int, prediction: Prophecy) -> None:
+        """Helper function for rendering assembly.
         """
 
-        opcode = "".join(f"{b:02x}" for b in insn.bytes)
+        def __render_asm_line(insn: InsnLike) -> str:
+            """Helper function for rendering assembly instructions, indicates where we are and
+            the branch prediction provided by branch predictor
+            """
+
+            trace_line = f"{insn.address:#010x} │ {insn.bytes.hex():18s} {insn.mnemonic:12} {insn.op_str:35s}"
+
+            cursor = ''  # current instruction cursor
+            brmark = ''  # branching mark
+
+            if insn.address == pc:
+                cursor = CURSOR
+
+                if prediction.going:
+                    # branch target might be None in case it should have been
+                    # read from memory but that memory could not be reached
+                    bmark = '?' if prediction.where is None else (GOING_DN if prediction.where > pc else GOING_UP)
+
+                    # apply some colors
+                    brmark = f'{color.RED}{bmark}{color.RESET}'
 
-        trace_line = f"0x{insn.address:08x} │ {opcode:15s} {insn.mnemonic:10} {insn.op_str:35s}"
+                # <DEBUG>
```

Comment:
maybe only show this messages when QL_VERBOSE is DEBUG ?

Reply:
yes it's good idea

Reply:
I'll figure something out.

Reply:
There's also another DEBUG message in `render.py`. Do you think thats also nessassary ?

Reply:
Snapshot diff prints removed.